ext/sockets/sockets.c

```diff
@@ -2054,25 +2065,33 @@ PHP_FUNCTION(socket_set_option)
 		}
 #ifdef SO_BINDTODEVICE
 		case SO_BINDTODEVICE: {
-			if (Z_TYPE_P(arg4) == IS_STRING) {
-				opt_ptr = Z_STRVAL_P(arg4);
-				optlen = Z_STRLEN_P(arg4);
-			} else {
-				opt_ptr = "";
-				optlen = 0;
+			if (Z_TYPE_P(arg4) != IS_STRING) {
+				zend_argument_type_error(4, "must be of type string when argument #3 ($option) is SO_BINDTODEVICE, %s given", zend_zval_value_name(arg4));
+				RETURN_THROWS();
 			}
+			opt_ptr = Z_STRVAL_P(arg4);
+			optlen = Z_STRLEN_P(arg4);
 			break;
 		}
 #endif
 
 #ifdef SO_ACCEPTFILTER
 		case SO_ACCEPTFILTER: {
 			if (Z_TYPE_P(arg4) != IS_STRING) {
-				php_error_docref(NULL, E_WARNING, "Invalid filter argument type");
-				RETURN_FALSE;
+				zend_argument_type_error(4, "must be of type string when argument #3 ($option) is SO_ACCEPTFILTER, %s given", zend_zval_value_name(arg4));
+				RETURN_THROWS();
+			}
+			if (zend_str_has_nul_byte(Z_STR_P(arg4))) {
+				zend_argument_value_error(4, "must not contain null bytes when argument #3 ($option) is SO_ACCEPTFILTER");
+				RETURN_THROWS();
 			}
 			struct accept_filter_arg af = {0};
-			strlcpy(af.af_name, Z_STRVAL_P(arg4), sizeof(af.af_name));
+			if (Z_STRLEN_P(arg4) > sizeof(af.af_name)) {
```

Comment:
Should that be `Z_STRLEN_P(arg4) >= sizeof(af.af_name)` (or `Z_STRLEN_P(arg4)+1 > sizeof(af.af_name)`) to account for the null byte?

Reply:
Like [here](https://github.com/php/php-src/pull/17185#issuecomment-2558462321), it would be nice to have more explanations of why you feel we need to replace strlcpy with memcpy.
1/ It is easier to fall into overflows with memcpy as per Arnaud/my comments.
2/ If you want to check truncations, strlcpy return value can help in this.
3/ If this is performance gain and since those changes are mainly about FreeBSD parts, let's say we create a bunch of sockets and we set bbr tcp stack for each so how much of a difference it makes in practice ? Albeit freebsd's memcpy finally [caught up a lot in term of optimisations (highlighting the case for size <= 32 bytes here)](https://github.com/freebsd/freebsd-src/blob/38b67578fb4bbf568f7012ca3921a4d15cfe7c5d/lib/libc/amd64/string/memmove.S#L106) since the 14.x release serie. In comparison [our strlcpy version is very straightforward](https://github.com/php/php-src/blob/0aeac680634e97a51a4ae1bbd4c9424f480ebb9a/main/strlcpy.c#L62) but today's compilers are smarter.

Whether the changes are justified, I do not think it is very urgent we can see after Xmas days no rush :)

Reply:
Because `strlcpy()` is bad as I just replied to that other thread, we also need to do extra work to instrumentalize this on top of using a non standard C function.

I don't mind adding a helper function if that's the concern.

Reply:
I see ... in this case you still to address arnaud and my comment.

Reply:
Yeah, I'll get round to it after the Xmas break :)